### PR TITLE
Fix resource consumption in reductions

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89144

Reductions along a (large enough) contiguous dimension vectorise the
loading of the inputs. This vectorisation was not taken into account
when computing the necessary resources for the kernel.

### CODE REVIEW

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 1]
```diff
@@ -1101,9 +1102,13 @@ ReduceConfig setReduceConfig(const TensorIterator& iter){
...
 
-  constexpr int min_values_per_thread = 16;
-  constexpr int max_values_per_thread = 256;
-  const int blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor / (block_width * block_height);
+  int min_values_per_thread = 16;
+  int max_values_per_thread = 256;
+  if (config.vectorize_input) {
+    min_values_per_thread *= config.input_vec_size;
+    max_values_per_thread *= config.input_vec_size;
+  }
```
Comment (lezcano) [inline aten/src/ATen/native/cuda/Reduce.cuh:1110]: This part I'm not 100% sure about. Is it alright to have larger limits of values per thread when we do vectorised loads, as 4 of them will be reduced sequentially by each thread after each load, or is it better to keep these max values per thread and then reorganise the logic in the next `if`?

Also, I just realised that we also have a `config.values_per_thread() >= 256` in the previous if, so if this fix of increasing the min/max values per thread is correct, we should move it up to the previous if as well and rewrite that as ` >= max_values_per_thread`.

Comment (lezcano) [reply aten/src/ATen/native/cuda/Reduce.cuh:1110]: update: increasing `max_values_per_thread` does no good. I'll keep looking into this

**Comment by lezcano** [comment]
So, after lots of thinking, this is as good as it gets. The issue here is that, for small contiguous vectors, for which the inputs are vectorised, the kernel was launching too many resources. e.g. for a vector of size 512, it was launching a block size of 512, and 3/4 of the threads were doing nothing.

After much thinking, this issue just happens for small vectors (those of length less than 2048). Thankfully, once you get to larger reductions, the computation is performed correctly, as the thread block remains constant of size 512, and the rest of the computations for the size of the grid are performed wrt. this value.

For this reason, even though this fix reduces the number of resources launched for some kernels, this only happens for tiny kernels, so we do not get any performance improvement :(

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 2]
```diff
@@ -1068,6 +1064,7 @@ ReduceConfig setReduceConfig(const TensorIterator& iter){
       // Note that if vt0 < ReduceConfig::vec_size, then this means the register pressure could be high, in such case,
       // we should avoid vectorization.
       config.vectorize_input = true;
+      dim0 /= config.input_vec_size;
```
Comment (lezcano) [inline aten/src/ATen/native/cuda/Reduce.cuh:1067]: This is the only relevant change of the PR, the rest are pretty much cosmetic changes.

**Comment by lezcano** [comment]
PTAL @zasdfgbnm

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 3]
```diff
@@ -589,9 +588,6 @@ struct ReduceOp {
       for (index_t i = 0; i < vt0; i++) {
         const auto offset = calc(idx + i * stride) / output_vec_size;
         values[i] = memory::load_vector<output_vec_size>(data_, offset);
-      }
-      #pragma unroll
-      for (index_t i = 0; i < vt0; i++) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Reduce.cuh:594]: Did you check if this change affected performance? 

Background: For pointwise ops, the two different ways of writing code leads to very different perf:
```CUDA
#pragma unroll
for (int i = 0; i < 4; i++) {
  load
  compute
  store
}
```
vs
```CUDA
#pragma unroll
for (int i = 0; i < 4; i++) {
  load
}
#pragma unroll
for (int i = 0; i < 4; i++) {
  compute
}
#pragma unroll
for (int i = 0; i < 4; i++) {
  store
}
```
I don't think we have the same issue here, but worth checking.

**Review by zasdfgbnm — APPROVED** [review]

**Comment by zasdfgbnm** [comment]
> For this reason, even though this fix reduces the number of resources launched for some kernels, this only happens for tiny kernels, so we do not get any performance improvement :(

Although there is no perf change, we still need to do the correct thing, so let's take this change.

File: aten/src/ATen/native/cuda/Reduce.cuh (hunk 3 above)
Comment (lezcano) [reply aten/src/ATen/native/cuda/Reduce.cuh]: That's actually very fair. Sorry for the sloppy edit. I just rolled it back to be on the safe side.

**Comment by lezcano** [comment]
@ngimel could you approve this one?

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**